warcprox/dedup.py

```diff
@@ -412,17 +415,30 @@ def _filter_and_bucketize(self, batch):
         '''
         buckets = collections.defaultdict(list)
         discards = []
+        # for duplicate checks, see https://webarchive.jira.com/browse/WT-31
+        hash_plus_urls = set()
         for recorded_url in batch:
+            if recorded_url.payload_digest:
+                hash_plus_url = b''.join((warcprox.digest_str(
+                            recorded_url.payload_digest, self.options.base32), recorded_url.url))
             if (recorded_url.response_recorder
                     and recorded_url.payload_digest
-                    and self.trough_dedup_db.should_dedup(recorded_url)):
+                    and self.trough_dedup_db.should_dedup(recorded_url)
+                    and hash_plus_url not in hash_plus_urls):
```

Comment:
Speed optimisation:
I would put `hash_plus_url not in has_plus_urls` before `self.trough_dedup_db.should_dedup(recorded_url)`
because it is much faster and you will avoid the slower though db lookup on duplicate records.